contracts/staking/BaseShareField.sol

```diff
@@ -142,11 +140,31 @@ contract BaseShareField is DAOContract{
         UserInfo storage userInfo = users[user];
         uint _accAmountPerShare = accAmountPerShare;
         // uint256 lpSupply = totalProductivity;
-        if (totalProductivity != 0) {
-            uint reward = _currentReward();
+        FundManager fm = FundManager(nameService.getAddress("FUND_MANAGER"));
+        uint pending = 0;
+        (uint rewardsPerBlock, uint blockStart, uint blockEnd, bool isBlackListed) = fm.getStakingReward(address(this));
+        if (totalProductivity != 0 && block.number >= blockStart && blockEnd>= block.number) {
+            uint256 multiplier = block.number.sub(lastRewardBlock);
+            uint256 reward = multiplier.mul(rewardsPerBlock * 10 ** 16); // turn it to 18 decimals
             _accAmountPerShare = _accAmountPerShare.add(reward.mul(1e12).div(totalProductivity));
+            uint256 blocksPaid = userInfo.lastRewardTime.sub(userInfo.multiplierResetTime);
+            uint256 blocksPassedFirstMonth = Math.min(172800,block.number.sub(userInfo.multiplierResetTime)); //172800 is equivalent of month in blocks 
+            uint256 blocksToPay = block.number.sub(userInfo.lastRewardTime);
+            uint256 firstMonthBlocksToPay = blocksPaid >= (172800) ? 0 : blocksPassedFirstMonth.sub(blocksPaid);
+            uint256 fullBlocksToPay = blocksToPay.sub(firstMonthBlocksToPay);
+            
+           
+            
+            if (blocksToPay != 0){
+                pending = userInfo.amount.mul(accAmountPerShare).div(1e12).sub(userInfo.rewardDebt);
+                uint rewardPerBlock = pending.mul(uint256(10 ** 12)).div(blocksToPay).div(uint256(10 ** 12)); // increase resolution
+                pending  = ((firstMonthBlocksToPay.mul(50*10**18).div(100)) + fullBlocksToPay).mul(rewardPerBlock).div(1e18); // divide 1e18 so reduce it to 18decimals
+              
+            }
+            
```

Comment:
seems like it requires DRY

Reply:
Yes few lines we can discuss if we can make it better

Reply:
this code repeats in both _takeWithAddress and _audit
where is _takeWithAddress used? what is the purpose of each? there are no comments on this method.
please go over all of the staking code and document ALL functions

Reply:
I implemented internal helper function